backend/server/db/helpers/users.py

```diff
@@ -0,0 +1,160 @@
+                "planner": "",
+            },
+            "$set": {
+                "setup": False,
+            },
+        },
+        upsert=False,
+        hint=UID_INDEX_NAME,
+    )
+
+    return res.matched_count == 1
+
+def set_user(uid: str, data: UserStorage, overwrite: bool = False) -> bool:
+    if overwrite:
+        # update/upsert
+        res = usersNewCOL.update_one(
+            { "uid": uid },
+            {
+                "$set": data.model_dump(include={ "degree", "courses", "planner", "setup" }),
+                "$setOnInsert": {
+                    # The fields that are usually immutable
+                    "uid": uid,
+                    "guest": data.guest,
+                },
+            },
+            upsert=True,
+            hint=UID_INDEX_NAME,
+        )
+
+        return res.matched_count == 1
+
+    # just try insert
+    try:
+        usersNewCOL.insert_one(data.model_dump()) # type: ignore
+        return True
+    except pymongo.errors.DuplicateKeyError:
+        # uid already existed
+        return False
+
+def update_user_degree(uid: str, data: UserDegreeStorage) -> bool:
+    res = usersNewCOL.update_one(
+        { "uid": uid, "setup": True },
+        {
+            "$set": {
+                "degree": data.model_dump(),
+            },
+        },
+        upsert=False,
+        hint=UID_INDEX_NAME,
+    )
+
+    return res.matched_count == 1
+
+def update_user_courses(uid: str, data: UserCoursesStorage) -> bool:
+    res = usersNewCOL.update_one(
+        { "uid": uid, "setup": True },
+        {
+            "$set": {
+                "courses": { code: info.model_dump() for code, info in data.items() },
+            },
+        },
+        upsert=False,
+        hint=UID_INDEX_NAME,
+    )
+
+    return res.matched_count == 1
+
+def update_user_planner(uid: str, data: UserPlannerStorage) -> bool:
+    res = usersNewCOL.update_one(
+        { "uid": uid, "setup": True },
+        {
+            "$set": {
+                "planner": data.model_dump(),
+            },
+        },
+        upsert=False,
+        hint=UID_INDEX_NAME,
+    )
+
+    return res.matched_count == 1
+
+def update_user(uid: str, data: PartialUserStorage) -> bool:
+    # updates certain properties of the user
+    # if enough are given, declares it as setup
+    payload = {
+        k: v
+        for k, v
+        in data.model_dump(
+            include={ "courses", "degree", "planner" }, 
+            exclude_unset=True,
+        ).items()
+        if v is not None  # cannot exclude_none since subclasses use None
+    }
+
+    if len(payload) == 0:
+        # most semantically correct
+        return user_is_setup(uid)
+
+    if "courses" in payload and "degree" in payload and "planner" in payload:
+        # enough to declare user as setup
+        payload["setup"] = True
+
+    res = usersNewCOL.update_one(
+        { "uid": uid, "setup": True } if "setup" not in payload else { "uid": uid },
```

Comment:
Is that ternary the right way around?
In the first if-statement, having the key means that the user is fully set up.
However, in the update, we insert the `setup: True` if we don't have the setup key in the payload

Reply:
It's kinda confusing, but this is the filter arg for the update (what is used to find the documents we want to update).

You are right, we add the "setup" key into the payload above if they have given us all the other keys required to make a user setup. 
SO, if they haven't given us all the properties we need (and the key is not present), we are only patching certain fields, thus we require our search to look for a document that already setup (or we will be missing fields). 
OTHERWISE, if they have given us all the keys we need, then we can apply the update to the user regardless if they are setup or not. 

So we are doing:
> If we do not have all keys, look for a match that is already setup, otherwise just look for a user

Maybe I can flip the ternary to make this a bit clearer?